numba/_helperlib.c

```diff
@@ -977,18 +847,122 @@ numba_do_raise(PyObject *exc_packed)
         traceback_add(function_name_str, filename_str, \
                       (int)PyLong_AsLong(lineno));
     }
+}
 
-    /* PyErr_SetObject incref's its arguments */
-    Py_XDECREF(value);
-    Py_XDECREF(type);
-    return 0;
+static
+int reraise_exc_is_none() {
+    /* Reraise */
+    PyThreadState *tstate = PyThreadState_GET();
+    PyObject *tb, *type, *value;
+#if (PY_MAJOR_VERSION >= 3) && (PY_MINOR_VERSION >= 7)
+    _PyErr_StackItem *tstate_exc = tstate->exc_info;
+#else
+    PyThreadState *tstate_exc = tstate;
+#endif
+    type = tstate_exc->exc_type;
+    value = tstate_exc->exc_value;
+    tb = tstate_exc->exc_traceback;
+    if (type == Py_None) {
+        PyErr_SetString(PyExc_RuntimeError,
+                        "No active exception to reraise");
+        return 0;
+    }
+    /* incref needed because PyErr_Restore DOES NOT */
+    Py_XINCREF(type);
+    Py_XINCREF(value);
+    Py_XINCREF(tb);
+    PyErr_Restore(type, value, tb);
+    return 1;
+}
 
-raise_error:
-    Py_XDECREF(value);
-    Py_XDECREF(type);
-    return 0;
+/*
+ * PyExceptionClass_Check(exc) must be True
+ * value can be NULL
+ */
+static
+int process_exception_class(PyObject *exc, PyObject *value) {
+    PyObject *type;
+    /* It is a class, type used here just as a tmp var */
+    type = PyObject_CallObject(exc, value);
+    if (type == NULL){
+        return 0;
+    }
+    if (!PyExceptionInstance_Check(type)) {
+        PyErr_SetString(PyExc_TypeError,
+                        "exceptions must derive from BaseException");
+        Py_DECREF(type);
+        return 0;
+    }
+    /* all ok, set type to the exc */
+    Py_DECREF(type);
+    type = exc;
+    PyErr_SetObject(type, value);
+    return 1;
+}
+
+/*
```

Comment:
Would be good to expand the comment to include information about what this function does.

Reply:
Done in https://github.com/numba/numba/pull/6017/commits/07415ebafbe9ae43e20e824762ce6e3941b4f21c